File: python-sdk/reeeductio/kvdata.py
"""
Simple key-value data store helpers for reeeductio.

Provides utilities for reading and writing signed data entries.
"""

from __future__ import annotations

from datetime import datetime, timezone

import httpx

from .crypto import encode_base64, sign_data
from .exceptions import NotFoundError, ValidationError
from .models import DataEntry


def compute_data_signature(
    space_id: str,
    path: str,
    data: bytes,
    signed_at: int,
    private_key: bytes,
) -> bytes:
    """
    Compute signature for data entry.

    Signature is over: space_id|path|base64(data)|signed_at

    Args:
        space_id: Typed space identifier
        path: Data path
        data: Data bytes
        signed_at: Unix timestamp in milliseconds
        private_key: Signer's Ed25519 private key

    Returns:
        64-byte signature
    """
    # Signature is over: space_id|path|base64(data)|signed_at
    # The server verifies against the base64-encoded data string from the JSON body,
    # so we must sign over the same base64 representation.
    data_b64 = encode_base64(data)
    sig_input = f"{space_id}|{path}|{data_b64}|{signed_at}".encode()
    return sign_data(sig_input, private_key)


def get_data(
    client: httpx.Client,
    space_id: str,
    path: str,
) -> DataEntry:
    """
    Get data value at path.

    Args:
        client: Authenticated httpx client
        space_id: Typed space identifier
        path: Data path (e.g., "profiles/alice", "settings/theme")

    Returns:
        DataEntry with the stored data

    Raises:
        NotFoundError: If no data exists at this path
        ValidationError: If request fails
    """
    try:
        response = client.get(f"/spaces/{space_id}/data/{path}")
        response.raise_for_status()
        data = response.json()
        return DataEntry(**data)
    except httpx.HTTPStatusError as e:
        if e.response.status_code == 404:
            raise NotFoundError(f"No data found at path: {path}") from e
        raise ValidationError(f"Failed to get data: {e.response.text}") from e
    except Exception as e:
        raise ValidationError(f"Failed to get data: {e}") from e


async def get_data_async(
    client: httpx.AsyncClient,
    space_id: str,
    path: str,
) -> DataEntry:
    """
    Async version of get_data.

    Args:
        client: Authenticated async httpx client
        space_id: Typed space identifier
        path: Data path

    Returns:
        DataEntry with the stored data

    Raises:
        NotFoundError: If no data exists at this path
        ValidationError: If request fails
    """
    try:
        response = await client.get(f"/spaces/{space_id}/data/{path}")
        response.raise_for_status()
        data = response.json()
        return DataEntry(**data)
    except httpx.HTTPStatusError as e:
        if e.response.status_code == 404:
            raise NotFoundError(f"No data found at path: {path}") from e
        raise ValidationError(f"Failed to get data: {e.response.text}") from e
    except Exception as e:
        raise ValidationError(f"Failed to get data: {e}") from e


def set_data(
    client: httpx.Client,
    space_id: str,
    path: str,
    data: bytes,
    signed_by: str,
    private_key: bytes,
) -> int:
    """
    Set data value at path with cryptographic signature.

    Args:
        client: Authenticated httpx client
        space_id: Typed space identifier
        path: Data path
        data: Data to store
        signed_by: Typed user/tool identifier of signer
        private_key: Signer's Ed25519 private key

    Returns:
        Timestamp when the data was signed (milliseconds)

    Raises:
        ValidationError: If request fails or signature is invalid
    """
    # Current timestamp in milliseconds
    signed_at = int(datetime.now(timezone.utc).timestamp() * 1000)

    # Compute signature
    signature = compute_data_signature(
        space_id=space_id,
        path=path,
        data=data,
        signed_at=signed_at,
        private_key=private_key,
    )

    # Create request body
    body = {
        "data": encode_base64(data),
        "signature": encode_base64(signature),
        "signed_by": signed_by,
        "signed_at": signed_at,
    }

    try:
        response = client.put(f"/spaces/{space_id}/data/{path}", json=body)
        response.raise_for_status()
        result = response.json()
        return result["signed_at"]
    except httpx.HTTPStatusError as e:
        raise ValidationError(f"Failed to set data: {e.response.text}") from e
    except Exception as e:
        raise ValidationError(f"Failed to set data: {e}") from e


async def set_data_async(
    client: httpx.AsyncClient,
    space_id: str,
    path: str,
    data: bytes,
    signed_by: str,
    private_key: bytes,
) -> int:
    """
    Async version of set_data.

    Args:
        client: Authenticated async httpx client
        space_id: Typed space identifier
        path: Data path
        data: Data to store
        signed_by: Typed user/tool identifier of signer
        private_key: Signer's Ed25519 private key

    Returns:
        Timestamp when the data was signed (milliseconds)

    Raises:
        ValidationError: If request fails or signature is invalid
    """
    # Current timestamp in milliseconds
    signed_at = int(datetime.now(timezone.utc).timestamp() * 1000)

    # Compute signature
    signature = compute_data_signature(
        space_id=space_id,
        path=path,
        data=data,
        signed_at=signed_at,
        private_key=private_key,
    )

    # Create request body
    body = {
        "data": encode_base64(data),
        "signature": encode_base64(signature),
        "signed_by": signed_by,
        "signed_at": signed_at,
    }

    try:
        response = await client.put(f"/spaces/{space_id}/data/{path}", json=body)
        response.raise_for_status()
        result = response.json()
        return result["signed_at"]
    except httpx.HTTPStatusError as e:
        raise ValidationError(f"Failed to set data: {e.response.text}") from e
    except Exception as e:
        raise ValidationError(f"Failed to set data: {e}") from e

File: typescript-sdk/src/kvdata.ts
/**
 * Simple key-value data store helpers for reeeductio.
 *
 * Provides utilities for reading and writing signed data entries.
 */

import {
  signData,
  encodeBase64,
  stringToBytes,
} from './crypto.js';
import { debugLog, errorLog } from './debug.js';
import type {
  DataEntry,
  DataSetResponse,
  ApiError,
} from './types.js';
import { createApiError, NotFoundError } from './exceptions.js';

/**
 * Compute signature for data entry.
 *
 * Signature is over: space_id|path|base64(data)|signed_at
 *
 * @param spaceId - Typed space identifier
 * @param path - Data path
 * @param data - Data bytes
 * @param signedAt - Unix timestamp in milliseconds
 * @param privateKey - Signer's Ed25519 private key
 * @returns 64-byte signature
 */
export async function computeDataSignature(
  spaceId: string,
  path: string,
  data: Uint8Array,
  signedAt: number,
  privateKey: Uint8Array
): Promise<Uint8Array> {
  // Signature is over: space_id|path|base64(data)|signed_at
  // The server verifies against the base64-encoded data string from the JSON body,
  // so we must sign over the same base64 representation.
  const dataB64 = encodeBase64(data);
  const sigInput = stringToBytes(`${spaceId}|${path}|${dataB64}|${signedAt}`);
  return signData(sigInput, privateKey);
}

/**
 * Get data value at path.
 *
 * @param fetchFn - Fetch function
 * @param baseUrl - API base URL
 * @param token - JWT bearer token
 * @param spaceId - Typed space identifier
 * @param path - Data path (e.g., "profiles/alice", "settings/theme")
 * @returns DataEntry with the stored data
 */
export async function getData(
  fetchFn: typeof fetch,
  baseUrl: string,
  token: string,
  spaceId: string,
  path: string
): Promise<DataEntry> {
  const url = `${baseUrl}/spaces/${spaceId}/data/${path}`;

  debugLog('data', 'GET data', { url, spaceId, path });
  const response = await fetchFn(url, {
    method: 'GET',
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    if (response.status === 404) {
      debugLog('data', 'GET data not found', { status: response.status, spaceId, path });
      throw new NotFoundError(`No data found at path: ${path}`);
    }
    const error = await parseError(response);
    errorLog('data', 'GET data failed', { status: response.status, error, spaceId, path });
    throw createApiError(response.status, error);
  }

  const entry = (await response.json()) as DataEntry;
  debugLog('data', 'GET data ok', { status: response.status, spaceId, path });
  return entry;
}

/**
 * Set data value at path with cryptographic signature.
 *
 * @param fetchFn - Fetch function
 * @param baseUrl - API base URL
 * @param token - JWT bearer token
 * @param spaceId - Typed space identifier
 * @param path - Data path
 * @param data - Data to store
 * @param signedBy - Typed user/tool identifier of signer
 * @param privateKey - Signer's Ed25519 private key
 * @returns DataSetResponse with the path and timestamp
 */
export async function setData(
  fetchFn: typeof fetch,
  baseUrl: string,
  token: string,
  spaceId: string,
  path: string,
  data: Uint8Array,
  signedBy: string,
  privateKey: Uint8Array
): Promise<DataSetResponse> {
  // Current timestamp in milliseconds
  const signedAt = Date.now();

  // Compute signature
  const signature = await computeDataSignature(
    spaceId,
    path,
    data,
    signedAt,
    privateKey
  );

  // Create request body
  const body = {
    data: encodeBase64(data),
    signature: encodeBase64(signature),
    signed_by: signedBy,
    signed_at: signedAt,
  };

  const url = `${baseUrl}/spaces/${spaceId}/data/${path}`;
  debugLog('data', 'PUT data', { url, spaceId, path, dataBytes: data.length, signedAt });
  const response = await fetchFn(url, {
    method: 'PUT',
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });

  if (!response.ok) {
    const error = await parseError(response);
    errorLog('data', 'PUT data failed', { status: response.status, error, spaceId, path });
    throw createApiError(response.status, error);
  }

  const result = (await response.json()) as DataSetResponse;
  debugLog('data', 'PUT data ok', { status: response.status, spaceId, path, signedAt: result.signed_at });
  return result;
}

/**
 * Parse error response from API.
 */
async function parseError(response: Response): Promise<ApiError | undefined> {
  try {
    return (await response.json()) as ApiError;
  } catch {
    return undefined;
  }
}
